### PR TITLE
Joachim Rayski - OXO

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
holy crap spent so much time trying to figure out why my code was failing with travis and the cause was me initializing my variables in the class itself instead of in __init__. Damn

File: oxo.py [hunk 1]
```diff
@@ -1,29 +1,159 @@
...
-        raise NotImplementedError("TODO: implement __init__")
+        self.board_rows = 3
+        self.board_columns = 3
+        self.board_array = []
+        # create the board array
+        for x in range(self.board_rows):
+            self.board_array.append([])
+
+            for y in range(self.board_columns):
+                self.board_array[x].append(0)
```
Comment (edpowley) [inline oxo.py:13]: Would be slightly more efficient to build the array first and then append it, e.g.:
```
        for x in range(self.board_rows):
            row = []
            for y in range(self.board_columns):
                row.append(0)
            self.board_array.append(row)
```
This cuts down on array accesses through `self.board_array[x]`

File: oxo.py [hunk 2]
```diff
@@ -1,29 +1,159 @@
...
+
+        self.show()
 
     def get_square(self, x, y):
         """ Return 0, 1 or 2 depending on the contents of the specified square. """
-        raise NotImplementedError("TODO: implement get_square")
+
+        square_content = self.board_array[x][y]
+        return square_content
```
Comment (edpowley) [inline oxo.py:21]: No real need for a temporary variable here -- could just `return self.board_array[x][y]`

File: oxo.py [hunk 3]
```diff
@@ -1,29 +1,159 @@
...
+        full_columns = 0
+
+        # check if all of the columns are full
+        for x in range(self.board_rows):
+            if 0 in self.board_array[x]:
+                return False
+            else:
+                full_columns += 1
+        if full_columns == self.board_rows:
+            return True
```
Comment (edpowley) [inline oxo.py:46]: Can there ever be a situation where the end of the `for` loop is reached but `full_columns` is not equal to `self.board_rows`? (If so then your function would not return a value, which would be a bug)

File: oxo.py [hunk 4]
```diff
@@ -1,29 +1,159 @@
...
+
+        for x in range(self.board_rows):
+
+            # check for player 1
+            if 0 not in self.board_array[x] and 2 not in self.board_array[x]:
+                return 1
+
+            # check for player 2
+            if 0 not in self.board_array[x] and 1 not in self.board_array[x]:
+                return 2
```
Comment (edpowley) [inline oxo.py:133]: Clever!

File: oxo.py [hunk 5]
```diff
@@ -1,29 +1,159 @@
...
+                if player2_score == self.board_columns:
+                    return 2
+
+        return 0
+
+    def check_diagonal(self):
+        """Check if a player has connected a row diagonally."""
+
+        connected_o_1 = 0
+        connected_x_1 = 0
```
Comment (edpowley) [inline oxo.py:75]: You have used `player1` and `player2` in `check_horizontal`, but `o` and `x` here. Better to be consistent

**Review by edpowley — COMMENTED** [review]
Good -- all tests passed. More comments woudl aid readability. Some capacity for expansion, though diagonal checks would need rethinking for non-square grids.

**Pull request closed without merging**